src/compiler/transformer.ts

```diff
@@ -393,6 +393,8 @@ namespace ts {
             lexicalEnvironmentFunctionDeclarations = undefined!;
             lexicalEnvironmentStatements = undefined!;
             lexicalEnvironmentFlags = LexicalEnvironmentFlags.None;
+            // Also start a block scope to allow declaration of block scope variables
+            startBlockScope();

```

Comment:
I think you're missing a call to `endBlockScope` in `endLexicalEnvironment`. It also might be a good idea to support suspending/resuming block scopes in `suspendLexicalEnvironment` and `resumeLexicalEnvironment`. Those are checks used to ensure we're in the correct lexical environment in between parsing parameter lists and function bodies, and that someone doesn't use a function like `visitParameterList` without pairing it with a call to `visitFunctionBody` (or applicable logic for resuming the lexical environment).

Reply:
I'm not sure this quite works the way you want, because we only ever call `startBlockScope` and `endBlockScope` from `visitIterationBody` currently, so we may not be including block scope declarations in the right places.

Reply:
I do call `endBlockScope` in `endLexicalEnviroment`.
```
            let statements: Statement[] | undefined = endBlockScope();
```

But looking at the code again there may be cases where `statements` gets overwritten. 

I will add support for resume and suspend. 

I did see that this was only used from an iteration block. The idea was to also create a new block scope when a lexical env is also created. If an iteration scope is encountered a new block scope is pushed and we will get let declarations added to the block. If we are not in an interation, the same scope in which `var` declarations would have previously been added gets used. 

There indeed are other block scopes that should be supported (and now that I think about it that is what you probably mean). Those don't really impact the runtime correctness of the current code that we currently generate but it would be a good idea to make `addBlockScopedVariable` function the way we expect in all circumstances. I will make this change shortly.